subprojects/core/src/integTest/groovy/org/gradle/api/provider/ProviderIntegrationTest.groovy

```diff
@@ -0,0 +1,381 @@
+                    enabled.get()
+                }
+                
+                void setEnabled(Provider<Boolean> enabled) {
+                    this.enabled = enabled
+                }
+                
+                @OutputFiles
+                FileCollection getOutputFiles() {
+                    outputFiles.get()
+                }
+
+                void setOutputFiles(Provider<FileCollection> outputFiles) {
+                    this.outputFiles = outputFiles
+                }
+                
+                @TaskAction
+                void resolveValue() {
+                    if (getEnabled()) {
+                        getOutputFiles().each {
+                            it.text = '$OUTPUT_FILE_CONTENT'
+                        }
+                    }
+                }
+            }
+        """
+
+        when:
+        succeeds('myTask')
+
+        then:
+        !defaultOutputFile.exists()
+    }
+
+    @Unroll
+    def "can lazily map extension property value to task property with #mechanism"() {
+        given:
+        buildFile << pluginWithExtensionMapping { taskConfiguration }
+        buildFile << customGroovyBasedTaskType()
+
+        when:
+        succeeds('myTask')
+
+        then:
+        !defaultOutputFile.exists()
+        customOutputFile.isFile()
+        customOutputFile.text == OUTPUT_FILE_CONTENT
+
+        where:
+        mechanism            | taskConfiguration
+        'convention mapping' | taskConfiguredWithConventionMapping()
+        'provider'           | taskConfiguredWithProvider()
+    }
+
+    def "can use provider to define a task dependency"() {
+        given:
+        buildFile << """
+            task producer {
+                ext.destFile = file('test.txt')
+                outputs.file destFile
+
+                doLast {
+                    destFile << '$OUTPUT_FILE_CONTENT'
+                }
+            }
+
+            def targetFile = provider { producer.destFile }
+            
+            targetFile.builtBy producer
+
+            task consumer {
+                dependsOn targetFile
+
+                doLast {
+                    println targetFile.get().text
+                }
+            }
+        """
+
+        when:
+        succeeds('consumer')
+
+        then:
+        executedTasks.containsAll(':producer', ':consumer')
+        outputContains(OUTPUT_FILE_CONTENT)
+    }
+
+    def "can use provider type to infer task dependency"() {
+        given:
+        buildFile << """
+            task producer(type: Producer) {
+                text = '$OUTPUT_FILE_CONTENT'
+                outputFiles = provider { files("\$buildDir/helloWorld.txt") }
+            }
+
+            task consumer(type: Consumer) {
+                inputFiles = producer.outputs.files
+            }
+
+            class Producer extends DefaultTask {
+                @Input
+                String text
+
+                Provider<FileCollection> outputFiles
```

Comment:
> If we'd go for it I believe we'd have to implement Provider for a FileCollection and FileTree separately

We have these already: `ConfigurableFileCollection` and `ConfigurableFileTree`.

> The creation of a FileCollection or FileTree would happen as usual (via the methods on Project). 

We should use the same pattern as we do for other properties. So if we inject a factory for say String properties, then we should inject a factory for file collections. Or if we inject into the fields we do the same.

> I am still not convinced that users would adopt the pattern that a FileCollection is-a Provider

I'm not sure what the actual problem is in your example.

A file collection _is_ a provider. This is the case now. It will still be the case in the future. I think it's much better to strongly type this so that there is in fact a single way of doing things.